File: edge-agent/src/wal.rs
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ActionType {
    #[serde(rename = "correction")]
    Correction,
    #[serde(rename = "decision")]
    Decision,
    #[serde(rename = "state_change")]
    StateChange,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub timestamp: String,
    pub agent_id: String,
    pub action_type: ActionType,
    pub payload: serde_json::Value,
    pub applied: bool,
}

/// Append-only write-ahead log for edge agent state
pub struct WAL {
    path: PathBuf,
    entries: Vec<Entry>,
}

impl WAL {
    pub fn open(dir: &Path) -> Result<Self, Box<dyn std::error::Error>> {
        fs::create_dir_all(dir)?;
        let path = dir.join("wal.json");
        let entries = if path.exists() {
            let data = fs::read_to_string(&path)?;
            serde_json::from_str(&data)?
        } else {
            Vec::new()
        };
        Ok(Self { path, entries })
    }

    pub fn append(
        &mut self,
        agent_id: &str,
        action: ActionType,
        payload: serde_json::Value,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let entry = Entry {
            timestamp: chrono_now(),
            agent_id: agent_id.to_string(),
            action_type: action,
            payload,
            applied: false,
        };
        self.entries.push(entry);
        self.persist()
    }

    pub fn mark_applied(&mut self, index: usize) -> Result<(), Box<dyn std::error::Error>> {
        let entry = self.entries.get_mut(index).ok_or("index out of range")?;
        entry.applied = true;
        self.persist()
    }

    pub fn unapplied(&self) -> Vec<&Entry> {
        self.entries.iter().filter(|e| !e.applied).collect()
    }

    pub fn unapplied_for_agent(&self, agent_id: &str) -> Vec<&Entry> {
        self.entries
            .iter()
            .filter(|e| !e.applied && e.agent_id == agent_id)
            .collect()
    }

    fn persist(&self) -> Result<(), Box<dyn std::error::Error>> {
        let data = serde_json::to_string_pretty(&self.entries)?;
        fs::write(&self.path, data)?;
        Ok(())
    }
}

fn chrono_now() -> String {
    // Simple UTC timestamp without pulling in chrono crate
    let dur = std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap_or_default();
    format!("{}Z", dur.as_secs())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_wal_append_and_replay() {
        let dir = tempdir().unwrap();
        let mut w = WAL::open(dir.path()).unwrap();

        w.append("a1", ActionType::Correction, serde_json::json!({"k": "v"}))
            .unwrap();
        w.append("a1", ActionType::Decision, serde_json::json!("x"))
            .unwrap();

        assert_eq!(w.unapplied().len(), 2);

        w.mark_applied(0).unwrap();
        assert_eq!(w.unapplied().len(), 1);

        // Reload
        let w2 = WAL::open(dir.path()).unwrap();
        assert_eq!(w2.unapplied().len(), 1);
    }
}

File: internal/wal/wal.go
package wal

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// ActionType categorizes WAL entries
type ActionType string

const (
	ActionCorrection  ActionType = "correction"
	ActionDecision    ActionType = "decision"
	ActionStateChange ActionType = "state_change"
)

// Entry represents a single WAL entry
type Entry struct {
	Timestamp time.Time       `json:"timestamp"`
	AgentID   string          `json:"agent_id"`
	Action    ActionType      `json:"action_type"`
	Payload   json.RawMessage `json:"payload"`
	Applied   bool            `json:"applied"`
}

// WAL is an append-only write-ahead log for agent state
type WAL struct {
	dir     string
	mu      sync.Mutex
	entries []Entry
}

// New creates or opens a WAL in the given directory
func New(dir string) (*WAL, error) {
	if err := os.MkdirAll(dir, 0750); err != nil {
		return nil, fmt.Errorf("create wal dir: %w", err)
	}
	w := &WAL{dir: dir}
	if err := w.load(); err != nil {
		return nil, fmt.Errorf("load wal: %w", err)
	}
	return w, nil
}

// Append writes an entry to the WAL before the agent acts
func (w *WAL) Append(agentID string, action ActionType, payload interface{}) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	raw, err := json.Marshal(payload)
	if err != nil {
		return fmt.Errorf("marshal payload: %w", err)
	}

	entry := Entry{
		Timestamp: time.Now().UTC(),
		AgentID:   agentID,
		Action:    action,
		Payload:   raw,
		Applied:   false,
	}
	w.entries = append(w.entries, entry)
	return w.persist()
}

// MarkApplied marks an entry as applied by index
func (w *WAL) MarkApplied(index int) error {
	w.mu.Lock()
	defer w.mu.Unlock()

	if index < 0 || index >= len(w.entries) {
		return fmt.Errorf("index %d out of range [0, %d)", index, len(w.entries))
	}
	w.entries[index].Applied = true
	return w.persist()
}

// Unapplied returns all entries that haven't been applied yet
func (w *WAL) Unapplied() []Entry {
	w.mu.Lock()
	defer w.mu.Unlock()

	var result []Entry
	for _, e := range w.entries {
		if !e.Applied {
			result = append(result, e)
		}
	}
	return result
}

// UnappliedForAgent returns unapplied entries for a specific agent
func (w *WAL) UnappliedForAgent(agentID string) []Entry {
	w.mu.Lock()
	defer w.mu.Unlock()

	var result []Entry
	for _, e := range w.entries {
		if !e.Applied && e.AgentID == agentID {
			result = append(result, e)
		}
	}
	return result
}

// Entries returns all entries (snapshot)
func (w *WAL) Entries() []Entry {
	w.mu.Lock()
	defer w.mu.Unlock()
	out := make([]Entry, len(w.entries))
	copy(out, w.entries)
	return out
}

// Len returns the number of entries
func (w *WAL) Len() int {
	w.mu.Lock()
	defer w.mu.Unlock()
	return len(w.entries)
}

func (w *WAL) walPath() string {
	return filepath.Join(w.dir, "wal.json")
}

func (w *WAL) persist() error {
	data, err := json.MarshalIndent(w.entries, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(w.walPath(), data, 0640)
}

func (w *WAL) load() error {
	data, err := os.ReadFile(w.walPath())
	if err != nil {
		if os.IsNotExist(err) {
			w.entries = nil
			return nil
		}
		return err
	}
	return json.Unmarshal(data, &w.entries)
}
